File: ai-cv-scanner/scripts/scan-setup.py
#!/usr/bin/env python3
"""
scans ~/.claude/ directory for evidence of advanced AI usage.
outputs structured findings about skills, commands, MCP servers, hooks, and config.
"""

import json
import os
from pathlib import Path

CLAUDE_DIR = Path.home() / ".claude"

def scan_skills():
    """list installed skills with descriptions."""
    skills_dir = CLAUDE_DIR / "skills"
    results = []
    if not skills_dir.exists():
        return results

    for skill_dir in sorted(skills_dir.iterdir()):
        if not skill_dir.is_dir() and not skill_dir.is_symlink():
            continue
        name = skill_dir.name
        skill_md = skill_dir / "SKILL.md" if skill_dir.is_dir() else None
        if skill_dir.is_symlink():
            target = os.readlink(skill_dir)
            results.append({"name": name, "type": "symlink", "target": target})
        elif skill_md and skill_md.exists():
            desc = ""
            with open(skill_md) as f:
                for line in f:
                    if line.startswith("description:"):
                        desc = line.split(":", 1)[1].strip()
                        break
            results.append({"name": name, "type": "custom", "description": desc})
    return results

def scan_commands():
    """list custom slash commands."""
    commands_dir = CLAUDE_DIR / "commands"
    results = []
    if not commands_dir.exists():
        return results

    for cmd_file in sorted(commands_dir.glob("*.md")):
        with open(cmd_file) as f:
            content = f.read()[:200]
        results.append({"name": cmd_file.stem, "preview": content})
    return results

def scan_mcp_servers():
    """find MCP server configurations."""
    results = []
    for settings_file in [
        CLAUDE_DIR / "settings.json",
        CLAUDE_DIR / "settings.local.json",
    ]:
        if not settings_file.exists():
            continue
        try:
            with open(settings_file) as f:
                data = json.load(f)
            mcps = data.get("mcpServers", {})
            for name, config in mcps.items():
                cmd = config.get("command", "")
                args = config.get("args", [])
                results.append({
                    "name": name,
                    "command": cmd,
                    "args": args[:3] if isinstance(args, list) else [],
                    "source": settings_file.name,
                })
        except Exception:
            pass

    # also check project-level settings
    for project_dir in (CLAUDE_DIR / "projects").iterdir():
        settings = project_dir / "settings.json"
        if settings.exists():
            try:
                with open(settings) as f:
                    data = json.load(f)
                mcps = data.get("mcpServers", {})
                for name, config in mcps.items():
                    results.append({
                        "name": name,
                        "command": config.get("command", ""),
                        "source": f"projects/{project_dir.name}",
                    })
            except Exception:
                pass
    return results

def scan_hooks():
    """find configured hooks."""
    results = []
    for settings_file in [
        CLAUDE_DIR / "settings.json",
        CLAUDE_DIR / "settings.local.json",
    ]:
        if not settings_file.exists():
            continue
        try:
            with open(settings_file) as f:
                data = json.load(f)
            hooks = data.get("hooks", {})
            for event, hook_list in hooks.items():
                if isinstance(hook_list, list):
                    for hook in hook_list:
                        results.append({
                            "event": event,
                            "command": hook.get("command", ""),
                            "source": settings_file.name,
                        })
        except Exception:
            pass
    return results

def scan_claude_md_files():
    """find CLAUDE.md files and their sizes."""
    results = []
    # global
    global_md = CLAUDE_DIR / "CLAUDE.md"
    if global_md.exists():
        results.append({
            "path": str(global_md),
            "size": global_md.stat().st_size,
            "type": "global",
        })

    # project-level
    for project_dir in (CLAUDE_DIR / "projects").iterdir():
        if not project_dir.is_dir():
            continue
        # check for memory files too
        memory_dir = project_dir / "memory"
        if memory_dir.exists():
            memory_files = list(memory_dir.glob("*.md"))
            if memory_files:
                results.append({
                    "path": str(memory_dir),
                    "type": "memory",
                    "file_count": len(memory_files),
                    "project": project_dir.name,
                })

    # repo-level CLAUDE.md files
    git_dir = Path.home() / "git"
    if git_dir.exists():
        for repo_dir in git_dir.iterdir():
            claude_md = repo_dir / "CLAUDE.md"
            if claude_md.exists():
                results.append({
                    "path": str(claude_md),
                    "size": claude_md.stat().st_size,
                    "type": "project",
                    "project": repo_dir.name,
                })
    return results

def scan_history_stats():
    """basic stats from history.jsonl."""
    history = CLAUDE_DIR / "history.jsonl"
    if not history.exists():
        return {}

    total = 0
    projects = set()
    earliest = None
    latest = None

    with open(history) as f:
        for line in f:
            try:
                obj = json.loads(line)
                total += 1
                projects.add(obj.get("project", "unknown"))
                ts = obj.get("timestamp")
                if ts:
                    if earliest is None or ts < earliest:
                        earliest = ts
                    if latest is None or ts > latest:
                        latest = ts
            except Exception:
                pass

    return {
        "total_prompts": total,
        "unique_projects": len(projects),
        "earliest_timestamp": earliest,
        "latest_timestamp": latest,
    }

if __name__ == "__main__":
    report = {
        "skills": scan_skills(),
        "commands": scan_commands(),
        "mcp_servers": scan_mcp_servers(),
        "hooks": scan_hooks(),
        "claude_md_files": scan_claude_md_files(),
        "history_stats": scan_history_stats(),
    }

    print(json.dumps(report, indent=2, ensure_ascii=False, default=str))

File: deprecated/ai-cv-scanner/scripts/scan-setup.mjs
#!/usr/bin/env node

/**
 * scans ~/.claude/ for evidence of advanced AI usage.
 */

import { existsSync, lstatSync, readFileSync, readdirSync, readlinkSync, statSync } from 'fs';
import { join } from 'path';
import { homedir } from 'os';
import { createReadStream } from 'fs';
import { createInterface } from 'readline';

const CLAUDE_DIR = join(homedir(), '.claude');

/** lists installed skills with descriptions. */
function scanSkills() {
  const skillsDir = join(CLAUDE_DIR, 'skills');
  const results = [];
  if (!existsSync(skillsDir)) return results;

  for (const name of readdirSync(skillsDir).sort()) {
    const skillDir = join(skillsDir, name);
    const info = lstatSync(skillDir);
    if (!info.isDirectory() && !info.isSymbolicLink()) continue;

    if (info.isSymbolicLink()) {
      results.push({ name, type: 'symlink', target: readlinkSync(skillDir) });
      continue;
    }

    const skillMd = join(skillDir, 'SKILL.md');
    if (!existsSync(skillMd)) continue;

    let description = '';
    for (const line of readFileSync(skillMd, 'utf-8').split('\n')) {
      if (line.startsWith('description:')) {
        description = line.split(':').slice(1).join(':').trim();
        break;
      }
    }
    results.push({ name, type: 'custom', description });
  }

  return results;
}

/** lists custom slash commands. */
function scanCommands() {
  const commandsDir = join(CLAUDE_DIR, 'commands');
  const results = [];
  if (!existsSync(commandsDir)) return results;

  for (const file of readdirSync(commandsDir).sort()) {
    if (!file.endsWith('.md')) continue;
    const filePath = join(commandsDir, file);
    results.push({
      name: file.replace(/\.md$/, ''),
      preview: readFileSync(filePath, 'utf-8').slice(0, 200),
    });
  }
  return results;
}

/** finds MCP server configurations. */
function scanMcpServers() {
  const results = [];
  for (const settingsFile of [
    join(CLAUDE_DIR, 'settings.json'),
    join(CLAUDE_DIR, 'settings.local.json'),
  ]) {
    if (!existsSync(settingsFile)) continue;
    try {
      const data = JSON.parse(readFileSync(settingsFile, 'utf-8'));
      const mcps = data.mcpServers || {};
      for (const [name, config] of Object.entries(mcps)) {
        results.push({
          name,
          command: config.command || '',
          args: Array.isArray(config.args) ? config.args.slice(0, 3) : [],
          source: settingsFile.split('/').pop(),
        });
      }
    } catch {}
  }

  const projectsDir = join(CLAUDE_DIR, 'projects');
  if (!existsSync(projectsDir)) return results;

  for (const project of readdirSync(projectsDir)) {
    const settingsFile = join(projectsDir, project, 'settings.json');
    if (!existsSync(settingsFile)) continue;
    try {
      const data = JSON.parse(readFileSync(settingsFile, 'utf-8'));
      const mcps = data.mcpServers || {};
      for (const [name, config] of Object.entries(mcps)) {
        results.push({
          name,
          command: config.command || '',
          source: `projects/${project}`,
        });
      }
    } catch {}
  }

  return results;
}

/** finds configured hooks. */
function scanHooks() {
  const results = [];
  for (const settingsFile of [
    join(CLAUDE_DIR, 'settings.json'),
    join(CLAUDE_DIR, 'settings.local.json'),
  ]) {
    if (!existsSync(settingsFile)) continue;
    try {
      const data = JSON.parse(readFileSync(settingsFile, 'utf-8'));
      const hooks = data.hooks || {};
      for (const [event, hookList] of Object.entries(hooks)) {
        if (!Array.isArray(hookList)) continue;
        for (const hook of hookList) {
          results.push({
            event,
            command: hook.command || '',
            source: settingsFile.split('/').pop(),
          });
        }
      }
    } catch {}
  }
  return results;
}

/** finds CLAUDE.md and memory files. */
function scanClaudeMdFiles() {
  const results = [];
  const globalMd = join(CLAUDE_DIR, 'CLAUDE.md');
  if (existsSync(globalMd)) {
    results.push({
      path: globalMd,
      size: statSync(globalMd).size,
      type: 'global',
    });
  }

  const projectsDir = join(CLAUDE_DIR, 'projects');
  if (existsSync(projectsDir)) {
    for (const project of readdirSync(projectsDir)) {
      const projectDir = join(projectsDir, project);
      if (!statSync(projectDir).isDirectory()) continue;

      const memoryDir = join(projectDir, 'memory');
      if (!existsSync(memoryDir)) continue;
      const memoryFiles = readdirSync(memoryDir).filter(file => file.endsWith('.md'));
      if (memoryFiles.length) {
        results.push({
          path: memoryDir,
          type: 'memory',
          file_count: memoryFiles.length,
          project,
        });
      }
    }
  }

  const gitDir = join(homedir(), 'git');
  if (existsSync(gitDir)) {
    for (const repo of readdirSync(gitDir)) {
      const claudeMd = join(gitDir, repo, 'CLAUDE.md');
      if (!existsSync(claudeMd)) continue;
      results.push({
        path: claudeMd,
        size: statSync(claudeMd).size,
        type: 'project',
        project: repo,
      });
    }
  }

  return results;
}

/** calculates basic stats from history.jsonl. */
async function scanHistoryStats() {
  const history = join(CLAUDE_DIR, 'history.jsonl');
  if (!existsSync(history)) return {};

  let total = 0;
  const projects = new Set();
  let earliest = null;
  let latest = null;

  const rl = createInterface({ input: createReadStream(history) });
  for await (const line of rl) {
    try {
      const entry = JSON.parse(line);
      total += 1;
      projects.add(entry.project || 'unknown');
      const ts = entry.timestamp;
      if (!ts) continue;
      if (earliest === null || ts < earliest) earliest = ts;
      if (latest === null || ts > latest) latest = ts;
    } catch {}
  }

  return {
    total_prompts: total,
    unique_projects: projects.size,
    earliest_timestamp: earliest,
    latest_timestamp: latest,
  };
}

async function main() {
  const report = {
    skills: scanSkills(),
    commands: scanCommands(),
    mcp_servers: scanMcpServers(),
    hooks: scanHooks(),
    claude_md_files: scanClaudeMdFiles(),
    history_stats: await scanHistoryStats(),
  };

  console.log(JSON.stringify(report, null, 2));
}

main().catch(err => {
  console.error('error:', err.message);
  process.exit(1);
});
